utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+///      the next sub-node which will have a length of length/2.
+///      (The delta has its own compact encoding.)
+///      Otherwise, skip the "jump" delta to the next sub-node
+///      which will have a length of length-length/2.
+///
+/// For more information:
+/// - [ICU4C UCharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UCharsTrie.html)
+/// - [ICU4J CharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html) API.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone)]
+pub struct Char16Trie<'data> {
+    /// An array of u16 containing the trie data.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub data: ZeroVec<'data, u16>,
+}
+
+/// This struct represents an iterator over a Char16Trie.
+pub struct Char16TrieIterator<'a> {
+    /// A reference to the Char16Trie data to iterate over.
+    trie: &'a [<u16 as zerovec::ule::AsULE>::ULE],
+    /// Index of next trie unit to read, or None if there are no more matches.
+    pos: Option<usize>,
+    /// Remaining length of a linear-match node, minus 1, or None if not in
+    /// such a node.
+    remaining_match_length: Option<usize>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum TrieResult {
+    /// The input unit(s) did not continue a matching string.
+    /// Once next() return TrieResult::NoMatch, all further calls to next()
+    /// will also return TrieResult::NoMatch.
+    NoMatch,
+    /// The input unit(s) matched a string but there is no value for the string
+    /// so far.  (It is a prefix of a longer string.)
+    NoValue,
+    /// The input unit(s) continued a matching string and there is a value for
+    /// the string so far. No further input byte/unit can continue a matching
+    /// string.
+    FinalValue(i32),
+    /// The input unit(s) continued a matching string and there is a value for
+    /// the string so far.  Another input byte/unit can continue a matching
+    /// string.
+    Intermediate(i32),
+}
+
+impl<'a> Char16TrieIterator<'a> {
+    /// Returns a new [`Char16Iterator`] backed by borrowed data for the `trie` array
+    pub fn new(trie: &'a [<u16 as zerovec::ule::AsULE>::ULE]) -> Self {
+        Self {
+            trie,
+            pos: Some(0),
+            remaining_match_length: None,
+        }
+    }
+
+    /// Traverses the trie from the current state for this input char.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_char16trie::char16trie::{Char16Trie, Char16TrieIterator, TrieResult};
+    /// use zerovec::ZeroVec;
+    ///
+    /// // A Char16Trie containing the ASCII characters 'a' and 'b'.
+    /// let trie_data = vec![48, 97, 176, 98, 32868];
+    /// let trie = Char16Trie {
+    ///     data: ZeroVec::from_slice(trie_data.as_slice()),
+    /// };
+    ///
+    /// let mut itor = Char16TrieIterator::new(trie.data.as_slice());
```

Comment:
As above, it should be possible to make this object directly from the array. To the user, there is no value in the intermediate type/object currently called Char16Trie.

Reply:
In my opinion, the primary way to construct this object should be a method on the Char16Trie. That said, it's harmless to also have an alternate constructor from the slice like @dminor has here.